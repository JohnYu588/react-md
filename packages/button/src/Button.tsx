import {
  RippleContainer,
  useElementInteraction,
  useHigherContrastChildren,
} from "@react-md/core";
import type { ButtonHTMLAttributes } from "react";
import { forwardRef } from "react";
import type { ButtonClassNameThemeOptions } from "./styles";
import { getButtonClassName } from "./styles";

export interface ButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    ButtonClassNameThemeOptions {
  /** @defaultValue `"button"` */
  type?: "button" | "reset" | "submit";
}

/**
 * @example
 * Simple Example
 * ```tsx
 * import { Button } from "@react-md/button";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Theme Example
 * ```tsx
 * import { Button } from "@react-md/button";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="primary"
 *       themeType="contained"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Icon Button Example
 * ```tsx
 * import { Button } from "@react-md/button";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="secondary"
 *       themeType="outline"
 *       buttonType="icon"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Text Button with icons
 * ```tsx
 * import { Button } from "@react-md/button";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 */
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(props, ref) {
    const {
      type = "button",
      disabled = false,
      theme = "clear",
      themeType = "flat",
      buttonType = "text",
      className,
      children: propChildren,
      onClick,
      onKeyDown,
      onKeyUp,
      onMouseDown,
      onMouseUp,
      onMouseLeave,
      onTouchStart,
      onTouchEnd,
      onTouchMove,
      ...remaining
    } = props;
    const isThemeDisabled = theme === "disabled";
    const { pressed, pressedClassName, rippleContainerProps, handlers } =
      useElementInteraction({
        onClick,
        onKeyDown,
        onKeyUp,
        onMouseDown,
        onMouseUp,
        onMouseLeave,
        onTouchStart,
        onTouchEnd,
        onTouchMove,
        disabled: disabled || isThemeDisabled,
      });

    const children = useHigherContrastChildren(propChildren);

    return (
      <button
        {...remaining}
        // when the theme is set to `"disabled"`, the event handlers should be
        // removed so that it behaves like a disabled button. you do not want to
        // actually set the `disabled` attribute since it will lose keyboard
        // focus. this is mostly for supporting circular progress bars within
        // buttons
        {...(isThemeDisabled ? {} : handlers)}
        aria-disabled={isThemeDisabled || undefined}
        ref={ref}
        type={type}
        className={getButtonClassName({
          theme,
          themeType,
          buttonType,
          disabled,
          pressed,
          pressedClassName,
          className,
        })}
      >
        {children}
        {rippleContainerProps && <RippleContainer {...rippleContainerProps} />}
      </button>
    );
  }
);
