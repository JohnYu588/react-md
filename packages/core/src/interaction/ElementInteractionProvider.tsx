"use client";
import type { ReactElement, ReactNode } from "react";
import { createContext, useContext, useMemo } from "react";
import type { ElementInteractionMode } from "./types.js";

export interface ElementInteractionContext {
  mode: ElementInteractionMode;
  higherContrast: boolean;
}

const context = createContext<ElementInteractionContext>({
  mode: "ripple",
  higherContrast: true,
});
context.displayName = "ElementInteraction";
const { Provider } = context;

/**
 * @remarks \@since 6.0.9
 * @internal
 */
export function useElementInteractionContext(): Readonly<ElementInteractionContext> {
  return useContext(context);
}

/** @remarks \@since 6.0.9 */
export interface ElementInteractionProviderProps {
  /** {@inheritDoc ElementInteractionMode} */
  mode?: ElementInteractionMode;
  /**
   * This should be set to `true` if the `$disable-higher-contrast-interactions`
   * SCSS variable is set to `false`.
   *
   * @defaultValue `false`
   */
  disableHigherContrast?: boolean;
  children: ReactNode;
}

/**
 * **Client Component**
 *
 * Note: This is automatically included by the {@link CoreProviders}
 * @remarks \@since 6.0.9
 */
export function ElementInteractionProvider(
  props: ElementInteractionProviderProps
): ReactElement {
  const { mode = "ripple", disableHigherContrast = false, children } = props;
  const value = useMemo<ElementInteractionContext>(
    () => ({
      mode,
      higherContrast: !disableHigherContrast,
    }),
    [disableHigherContrast, mode]
  );

  return <Provider value={value}>{children}</Provider>;
}
