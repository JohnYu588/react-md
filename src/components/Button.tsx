import { useElementInteraction } from "@react-md/core";
import { cnb } from "cnbuilder";
import type { ButtonHTMLAttributes, ReactElement, ReactNode } from "react";

import styles from "./Button.module.scss";

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  buttonType?: "icon" | "text";
  children: ReactNode;
}

export function Button(props: ButtonProps): ReactElement {
  const {
    type = "button",
    disabled = false,
    className,
    buttonType = "text",
    children,
    ...remaining
  } = props;
  const { pressed, ...handlers } = useElementInteraction({
    disabled,
    ...remaining,
  });

  return (
    <button
      {...remaining}
      {...handlers}
      type={type}
      className={cnb(
        styles.button,
        buttonType === "text" && styles.text,
        buttonType === "icon" && styles.icon,
        pressed && styles.pressed,
        className
      )}
      disabled={disabled}
    >
      {children}
    </button>
  );
}
