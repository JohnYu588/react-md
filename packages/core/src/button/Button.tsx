"use client";
import { cnb } from "cnbuilder";
import type { ButtonHTMLAttributes } from "react";
import { forwardRef } from "react";
import { RippleContainer } from "../interaction/RippleContainer";
import { useElementInteraction } from "../interaction/useElementInteraction";
import { useHigherContrastChildren } from "../interaction/useHigherContrastChildren";
import type { PropsWithRef } from "../types";
import { bem } from "../utils";
import type {
  FloatingActionButtonPosition,
  FloatingActionButtonProps,
} from "./FloatingActionButton";
import { FloatingActionButton } from "./FloatingActionButton";

declare module "react" {
  interface CSSProperties {
    "--rmd-button-background-color"?: string;
    "--rmd-button-color"?: string;
    "--rmd-button-border-radius"?: string | number;
    "--rmd-button-contained-background-color"?: string;
    "--rmd-button-contained-color"?: string;
    "--rmd-button-text-horizontal-padding"?: string | number;
    "--rmd-button-text-vertical-padding"?: string | number;
    "--rmd-button-text-min-height"?: string | number;
    "--rmd-button-text-min-width"?: string | number;
    "--rmd-button-icon-size"?: string | number;
    "--rmd-button-icon-radius"?: string | number;
  }
}

const styles = bem("rmd-button");

/**
 * When this is set to `"text"`, the size of the button will be determined by
 * the content and will be more block-like. Icons can still be rendered in
 * `"text"` buttons and will have spacing automatically applied between other
 * content in the button.
 *
 * When this is set to `"icon"`, the button will be equal height/width and
 * circular.
 */
export type ButtonType = "text" | "icon" | "icon-square";

/**
 * One of the valid material design default button themes that can be used. This
 * will update the general look and feel by updating the colors within the
 * button while the `ButtonThemeType` will update the borders or box shadow.
 */
export type ButtonTheme =
  | "clear"
  | "primary"
  | "secondary"
  | "warning"
  | "error"
  | "disabled";

/**
 * When this is set to `"flat"`, the button will have no `background-color`,
 * `border`, and `box-shadow`. It will only set the `color` to the
 * {@link ButtonTheme}.
 *
 * When this is set to `"outline"`, the button will have no `background-color`,
 * but gain a `border` and `color` set to the {@link ButtonTheme}.
 *
 * When this is set to `"contained"`, the button will set the `background-color`
 * to the {@link ButtonTheme}, add some `box-shadow`, and set the `color` to
 * either `#000` or `#fff`. (The `color` defaults to whichever value has the
 * highest contrast ratio with the `background-color`)
 */
export type ButtonThemeType = "flat" | "outline" | "contained";

/** @remarks \@since 6.0.0 */
export interface ButtonClassNameThemeOptions {
  className?: string;

  /** @defaultValue `false` */
  disabled?: boolean;

  /**
   * @see {@link ButtonTheme} for information about the different types.
   * @defaultValue `"text"`
   */
  buttonType?: ButtonType;

  /** @defaultValue `"clear"` */
  theme?: ButtonTheme;

  /**
   * @see {@link ButtonThemeType} for information about the theming behavior.
   * @defaultValue `"flat"`
   */
  themeType?: ButtonThemeType;
}

/**
 * @remarks \@since 6.0.0
 * @internal
 */
export interface ButtonClassNameOptions extends ButtonClassNameThemeOptions {
  /** @defaultValue `false` */
  pressed?: boolean;
  pressedClassName?: string;
}

/**
 * Creates a button theme based on the button theming props. This is really just
 * used so that other elements like clickable `<div>`s or `<input type="file">`
 * can look like buttons.
 *
 * @param options - An object containing the themeable button props to generate a
 * button theme className.
 * @returns a string of class names to create an element with a button theme.
 * @remarks \@since 6.0.0 This used to be called `buttonThemeClassNames`.
 */
export function button(options: ButtonClassNameOptions = {}): string {
  const {
    theme: propTheme = "clear",
    themeType = "flat",
    buttonType = "text",
    disabled: propDisabled = false,
    pressed = false,
    pressedClassName,
    className,
  } = options;

  const theme = propTheme === "disabled" ? "clear" : propTheme;
  const disabled = propDisabled || propTheme === "disabled";
  const text = buttonType === "text";
  const icon = !text;
  const outline = themeType === "outline";
  const contained = themeType === "contained";
  const clear = theme === "clear";

  return cnb(
    styles({
      text,
      icon,
      "icon-square": buttonType === "icon-square",
      disabled,
      contained: !disabled && contained,
      outline,
      pressed: contained && pressed,
      [theme]: !disabled && !clear && contained,
      [`text-${theme}`]: !disabled && !clear && !contained,
      [`outline-${theme}`]: !disabled && !clear && outline,
    }),
    pressedClassName,
    className
  );
}

export interface ButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    ButtonClassNameThemeOptions {
  /** @defaultValue `"button"` */
  type?: "button" | "reset" | "submit";

  /**
   * The position within the viewport to display the button as a floating action
   * button.
   */
  floating?: FloatingActionButtonPosition;

  /**
   * Any additional props to provide the to `FAB` container element when the
   * `floating` prop is provided
   */
  floatingProps?: PropsWithRef<FloatingActionButtonProps, HTMLSpanElement>;
}

/**
 * **Client Component**
 *
 * @example
 * Simple Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Theme Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="primary"
 *       themeType="contained"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Icon Button Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="secondary"
 *       themeType="outline"
 *       buttonType="icon"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Text Button with icons
 * ```tsx
 * import { Button } from "@react-md/core";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 */
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(props, ref) {
    const {
      type = "button",
      disabled = false,
      floating = null,
      floatingProps,
      theme = floating ? "secondary" : "clear",
      themeType = floating ? "contained" : "flat",
      buttonType = floating ? "icon" : "text",
      className,
      children: propChildren,
      onBlur,
      onClick,
      onKeyDown,
      onKeyUp,
      onMouseDown,
      onMouseUp,
      onMouseLeave,
      onDragStart,
      onTouchStart,
      onTouchEnd,
      onTouchMove,
      ...remaining
    } = props;
    const isThemeDisabled = theme === "disabled";
    const { pressed, pressedClassName, rippleContainerProps, handlers } =
      useElementInteraction({
        onBlur,
        onClick,
        onKeyDown,
        onKeyUp,
        onMouseDown,
        onMouseUp,
        onMouseLeave,
        onDragStart,
        onTouchStart,
        onTouchEnd,
        onTouchMove,
        disabled: disabled || isThemeDisabled,
      });

    const children = useHigherContrastChildren(propChildren);

    return (
      <FloatingActionButton position={floating} {...floatingProps}>
        <button
          {...remaining}
          // when the theme is set to `"disabled"`, the event handlers should be
          // removed so that it behaves like a disabled button. you do not want to
          // actually set the `disabled` attribute since it will lose keyboard
          // focus. this is mostly for supporting circular progress bars within
          // buttons
          {...(isThemeDisabled ? {} : handlers)}
          aria-disabled={isThemeDisabled || remaining["aria-disabled"]}
          disabled={disabled}
          ref={ref}
          type={type}
          className={button({
            theme,
            themeType,
            buttonType,
            disabled,
            pressed,
            pressedClassName,
            className,
          })}
        >
          {children}
          {rippleContainerProps && (
            <RippleContainer {...rippleContainerProps} />
          )}
        </button>
      </FloatingActionButton>
    );
  }
);
