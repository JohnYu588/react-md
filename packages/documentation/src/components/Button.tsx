import type { ReactElement, ReactNode } from "react";
import type { ButtonProps } from "@react-md/button";
import { Button as RMDButton } from "@react-md/button";
import { Tooltip, useTooltip } from "@react-md/tooltip";

interface Props extends ButtonProps {
  id: string;
  tooltip?: ReactNode;
}

export default function Button({
  id,
  tooltip,
  children,
  onClick,
  onBlur,
  onFocus,
  onKeyDown,
  onMouseEnter,
  onMouseLeave,
  onTouchStart,
  onContextMenu,
  ...props
}: Props): ReactElement {
  const handlers = {
    onClick,
    onBlur,
    onFocus,
    onKeyDown,
    onMouseEnter,
    onMouseLeave,
    onTouchStart,
    onContextMenu,
  };
  const { elementProps, tooltipProps } = useTooltip({
    baseId: id,
    ...handlers,
  });

  return (
    <>
      <RMDButton {...props} {...(tooltip ? elementProps : { id, ...handlers })}>
        {children}
      </RMDButton>
      <Tooltip {...tooltipProps}>{tooltip}</Tooltip>
    </>
  );
}
