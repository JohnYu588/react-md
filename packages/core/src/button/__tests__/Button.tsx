import { describe, expect, it, jest } from "@jest/globals";
import { createRef } from "react";
import { ElementInteractionProvider } from "../../interaction/ElementInteractionProvider.js";
import { fireEvent, render } from "../../test-utils/index.js";
import { Button } from "../Button.js";
import {
  type ButtonTheme,
  type ButtonThemeType,
  type ButtonType,
} from "../buttonStyles.js";

describe("Button", () => {
  it("should apply the correct styling, HTML attributes, and allow a ref", () => {
    const ref = createRef<HTMLButtonElement>();
    const props = {
      ref,
      children: "Content",
    } as const;
    const { getByRole, rerender } = render(<Button {...props} />);

    const button = getByRole("button", { name: "Content" });
    expect(ref.current).toBeInstanceOf(HTMLButtonElement);
    expect(ref.current).toBe(button);
    expect(button).toMatchSnapshot();

    rerender(
      <Button
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(button).toMatchSnapshot();
  });

  it("should wrap text children in span elements to enable higher contrast unless the higher contrast is disabled", () => {
    const { getByRole, rerender } = render(<Button>Content</Button>, {
      wrapper: ({ children }) => (
        <ElementInteractionProvider disableHigherContrast>
          {children}
        </ElementInteractionProvider>
      ),
    });

    const button = getByRole("button", { name: "Content" });
    expect(button).toMatchSnapshot();

    rerender(
      <Button>
        <strong>Strong Content</strong>
      </Button>
    );
    expect(button).toMatchSnapshot();
  });

  it("should allow for the different themes", () => {
    const buttonTypes: readonly ButtonType[] = ["text", "icon"];
    const themes: readonly ButtonTheme[] = [
      "clear",
      "primary",
      "secondary",
      "warning",
      "error",
      "disabled",
    ];
    const themeTypes: readonly ButtonThemeType[] = [
      "flat",
      "outline",
      "contained",
    ];

    const { getByRole, rerender } = render(<Button>Content</Button>);
    const button = getByRole("button");
    expect(button).toMatchSnapshot();

    buttonTypes.forEach((buttonType) => {
      themes.forEach((theme) => {
        themeTypes.forEach((themeType) => {
          rerender(
            <Button theme={theme} themeType={themeType} buttonType={buttonType}>
              Content
            </Button>
          );

          expect(button).toMatchSnapshot();
        });
      });
    });
  });

  it("should not fire the onClick event handler when the theme is set to disabled", () => {
    const onClick = jest.fn();
    const { getByRole } = render(
      <Button onClick={onClick} theme="disabled">
        Button
      </Button>
    );
    const button = getByRole("button");

    fireEvent.click(button);
    expect(onClick).not.toHaveBeenCalled();
  });

  describe("floating action buttons", () => {
    it("should default to a secondary themed contained icon button if the floating behavior is enabled", () => {
      const { getByRole, rerender } = render(
        <Button aria-label="Button" floating="bottom-right" />
      );

      const button = getByRole("button", { name: "Button" });
      expect(button).toHaveClass("rmd-button--icon");
      expect(button.parentElement).toHaveClass("rmd-fab");
      expect(button.parentElement).toMatchSnapshot();

      rerender(<Button aria-label="Button" floating="top-right" />);
      expect(button).toHaveClass("rmd-button--icon");
      expect(button.parentElement).toHaveClass("rmd-fab");
      expect(button.parentElement).toMatchSnapshot();

      rerender(
        <Button
          aria-label="Button"
          floating="top-left"
          theme="primary"
          themeType="outline"
        />
      );
      expect(button).toHaveClass("rmd-button--icon");
      expect(button.parentElement).toHaveClass("rmd-fab");
      expect(button.parentElement).toMatchSnapshot();

      rerender(
        <Button
          floating="top-left"
          theme="clear"
          themeType="outline"
          buttonType="text"
        >
          Content
        </Button>
      );
      expect(button).toHaveClass("rmd-button--text");
      expect(button.parentElement).toHaveClass("rmd-fab");
      expect(button.parentElement).toMatchSnapshot();
    });

    it("should allow for custom props to be passed to the wrapper span element", () => {
      const ref = createRef<HTMLSpanElement>();
      const { getByRole } = render(
        <Button
          aria-label="Button"
          floating="bottom-left"
          floatingProps={{
            ref,
            style: { zIndex: 20 },
            className: "custom-class-name",
          }}
        />
      );

      const button = getByRole("button", { name: "Button" });
      const fab = button.parentElement;
      expect(ref.current).toBeInstanceOf(HTMLSpanElement);
      expect(ref.current).toBe(fab);
      expect(fab).toMatchSnapshot();
    });
  });
});
