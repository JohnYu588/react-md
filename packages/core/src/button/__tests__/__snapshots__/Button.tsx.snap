// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button floating action buttons should allow for custom props to be passed to the wrapper span element 1`] = `
<span
  class="rmd-fab rmd-fab--bl custom-class-name"
  style="z-index: 20;"
>
  <button
    aria-label="Button"
    class="rmd-button rmd-button--icon rmd-button--contained rmd-button--secondary"
    type="button"
  >
    <span
      class="rmd-ripple-container"
    />
  </button>
</span>
`;

exports[`Button floating action buttons should default to a secondary themed contained icon button if the floating behavior is enabled 1`] = `
<span
  class="rmd-fab rmd-fab--br"
>
  <button
    aria-label="Button"
    class="rmd-button rmd-button--icon rmd-button--contained rmd-button--secondary"
    type="button"
  >
    <span
      class="rmd-ripple-container"
    />
  </button>
</span>
`;

exports[`Button floating action buttons should default to a secondary themed contained icon button if the floating behavior is enabled 2`] = `
<span
  class="rmd-fab rmd-fab--tr"
>
  <button
    aria-label="Button"
    class="rmd-button rmd-button--icon rmd-button--contained rmd-button--secondary"
    type="button"
  >
    <span
      class="rmd-ripple-container"
    />
  </button>
</span>
`;

exports[`Button floating action buttons should default to a secondary themed contained icon button if the floating behavior is enabled 3`] = `
<span
  class="rmd-fab rmd-fab--tl"
>
  <button
    aria-label="Button"
    class="rmd-button rmd-button--icon rmd-button--outline rmd-button--text-primary rmd-button--outline-primary"
    type="button"
  >
    <span
      class="rmd-ripple-container"
    />
  </button>
</span>
`;

exports[`Button floating action buttons should default to a secondary themed contained icon button if the floating behavior is enabled 4`] = `
<span
  class="rmd-fab rmd-fab--tl"
>
  <button
    class="rmd-button rmd-button--text rmd-button--outline"
    type="button"
  >
    <span>
      Content
    </span>
    <span
      class="rmd-ripple-container"
    />
  </button>
</span>
`;

exports[`Button should allow for the different themes 1`] = `
<button
  class="rmd-button rmd-button--text"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 2`] = `
<button
  class="rmd-button rmd-button--text"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 3`] = `
<button
  class="rmd-button rmd-button--text rmd-button--outline"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 4`] = `
<button
  class="rmd-button rmd-button--text rmd-button--contained"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 5`] = `
<button
  class="rmd-button rmd-button--text rmd-button--text-primary"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 6`] = `
<button
  class="rmd-button rmd-button--text rmd-button--outline rmd-button--text-primary rmd-button--outline-primary"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 7`] = `
<button
  class="rmd-button rmd-button--text rmd-button--contained rmd-button--primary"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 8`] = `
<button
  class="rmd-button rmd-button--text rmd-button--text-secondary"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 9`] = `
<button
  class="rmd-button rmd-button--text rmd-button--outline rmd-button--text-secondary rmd-button--outline-secondary"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 10`] = `
<button
  class="rmd-button rmd-button--text rmd-button--contained rmd-button--secondary"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 11`] = `
<button
  class="rmd-button rmd-button--text rmd-button--text-warning"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 12`] = `
<button
  class="rmd-button rmd-button--text rmd-button--outline rmd-button--text-warning rmd-button--outline-warning"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 13`] = `
<button
  class="rmd-button rmd-button--text rmd-button--contained rmd-button--warning"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 14`] = `
<button
  class="rmd-button rmd-button--text rmd-button--text-error"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 15`] = `
<button
  class="rmd-button rmd-button--text rmd-button--outline rmd-button--text-error rmd-button--outline-error"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 16`] = `
<button
  class="rmd-button rmd-button--text rmd-button--contained rmd-button--error"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 17`] = `
<button
  aria-disabled="true"
  class="rmd-button rmd-button--text rmd-button--disabled"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 18`] = `
<button
  aria-disabled="true"
  class="rmd-button rmd-button--text rmd-button--disabled rmd-button--outline"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 19`] = `
<button
  aria-disabled="true"
  class="rmd-button rmd-button--text rmd-button--disabled"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 20`] = `
<button
  class="rmd-button rmd-button--icon"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 21`] = `
<button
  class="rmd-button rmd-button--icon rmd-button--outline"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 22`] = `
<button
  class="rmd-button rmd-button--icon rmd-button--contained"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 23`] = `
<button
  class="rmd-button rmd-button--icon rmd-button--text-primary"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 24`] = `
<button
  class="rmd-button rmd-button--icon rmd-button--outline rmd-button--text-primary rmd-button--outline-primary"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 25`] = `
<button
  class="rmd-button rmd-button--icon rmd-button--contained rmd-button--primary"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 26`] = `
<button
  class="rmd-button rmd-button--icon rmd-button--text-secondary"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 27`] = `
<button
  class="rmd-button rmd-button--icon rmd-button--outline rmd-button--text-secondary rmd-button--outline-secondary"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 28`] = `
<button
  class="rmd-button rmd-button--icon rmd-button--contained rmd-button--secondary"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 29`] = `
<button
  class="rmd-button rmd-button--icon rmd-button--text-warning"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 30`] = `
<button
  class="rmd-button rmd-button--icon rmd-button--outline rmd-button--text-warning rmd-button--outline-warning"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 31`] = `
<button
  class="rmd-button rmd-button--icon rmd-button--contained rmd-button--warning"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 32`] = `
<button
  class="rmd-button rmd-button--icon rmd-button--text-error"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 33`] = `
<button
  class="rmd-button rmd-button--icon rmd-button--outline rmd-button--text-error rmd-button--outline-error"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 34`] = `
<button
  class="rmd-button rmd-button--icon rmd-button--contained rmd-button--error"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 35`] = `
<button
  aria-disabled="true"
  class="rmd-button rmd-button--icon rmd-button--disabled"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 36`] = `
<button
  aria-disabled="true"
  class="rmd-button rmd-button--icon rmd-button--disabled rmd-button--outline"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should allow for the different themes 37`] = `
<button
  aria-disabled="true"
  class="rmd-button rmd-button--icon rmd-button--disabled"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should apply the correct styling, HTML attributes, and allow a ref 1`] = `
<button
  class="rmd-button rmd-button--text"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should apply the correct styling, HTML attributes, and allow a ref 2`] = `
<button
  class="rmd-button rmd-button--text custom-class-name"
  style="color: white;"
  type="button"
>
  <span>
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should wrap text children in span elements to enable higher contrast unless the higher contrast is disabled 1`] = `
<button
  class="rmd-button rmd-button--text"
  type="button"
>
  Content
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button should wrap text children in span elements to enable higher contrast unless the higher contrast is disabled 2`] = `
<button
  class="rmd-button rmd-button--text"
  type="button"
>
  <strong>
    Strong Content
  </strong>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Button styling utility class should be callable without any arguments 1`] = `"rmd-button rmd-button--text"`;
